File: src/wpcom.ts
export interface WPcomParams {
  fields?: string;
  parentId?: number;
  path: string;
  search?: string;
  site: string;
  slug?: string;
  type?: string;
}

export async function wpcomFetch(params: WPcomParams) {
  const { fields, parentId, path, search, site, slug, type } = params;
  const encodedFields = fields && `fields=${encodeURIComponent(fields)}`;
  let url = `https://public-api.wordpress.com/rest/v1.1/sites/${site}${path.replace(
    /\/$/,
    ''
  )}`;

  switch (true) {
    case !!slug:
      url = `${url}/slug:${params.slug}${fields && '?' + encodedFields}`;
      break;
    case !!parentId && !!type:
      url = `${url}/?type=${type}&parent_id=${params.parentId}${fields &&
        '&' + encodedFields}`;
      break;
    case !!type:
      url = `${url}/?type=${params.type}${fields && '&' + encodedFields}`;
      break;
    case !!search:
      url = `${url}/?search=${params.search}${fields && '&' + encodedFields}`;
      break;
    case !!fields:
      url = `${url}${fields && '?' + encodedFields}`;
      break;
    default:
      break;
  }

  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw Error(response.statusText);
    } else {
      return response.json();
    }
  } catch (error) {
    return error;
  }
}

export function wpcomGetThumbnailUrl(url: string, params: {}) {
  url = url.replace(/\/$/, '');
  if (params) {
    url = `${url}?${new URLSearchParams(params)}`;
  }
  return url;
}

File: web/lib/wpcom.dart
import 'dart:convert';
import 'dart:html';
import 'package:meta/meta.dart';

class WPcom {
  WPcom({@required this.site}) {
    baseUrl = 'https://public-api.wordpress.com/wp/v2/sites/$site';
  }
  final String site;
  String baseUrl;

  Future<List<Object>> getPosts() => makeGetRequest('$baseUrl/posts');

  Future<List<Object>> getCategories() => makeGetRequest('$baseUrl/categories');

  Future<List<Object>> getCategoryPosts(int id) =>
      makeGetRequest('$baseUrl/posts?categories=$id');

  Future<List<Object>> getTags() => makeGetRequest('$baseUrl/tags');

  Future<List<Object>> getTagPosts(int id) =>
      makeGetRequest('$baseUrl/posts?tags=$id');

  Future<List<Object>> search(String term) =>
      makeGetRequest('$baseUrl/search?search=$term');

  Future<List<Object>> makeGetRequest(String url) async {
    try {
      final response = await HttpRequest.getString(url);
      return jsonDecode(response) as List<Object>;
    } on Exception catch (error) {
      return [
        {
          'error': error.toString(),
        }
      ];
    }
  }
}

class WPcomParams {
  WPcomParams(
      {@required this.blog,
      this.fields,
      this.path,
      this.parentId,
      this.slug,
      this.type});
  final String blog;
  String fields;
  String path;
  int parentId;
  String slug;
  String type;
}
